netty/src/main/java/io/grpc/transport/netty/AbstractWriteQueue.java

```diff
@@ -0,0 +1,180 @@
+import javax.annotation.Nullable;
+
+/**
+ * A queue of pending writes to a {@link Channel} that is flushed as a single unit.
+ */
+abstract class AbstractWriteQueue {
+
+  private static final int DEQUE_CHUNK_SIZE = 128;
+
+  /**
+   * {@link Runnable} used to schedule work onto the tail of the event loop.
+   */
+  private final Runnable later = new Runnable() {
+    @Override
+    public void run() {
+      channel.write(AbstractWriteQueue.this);
+    }
+  };
+
+  private final Channel channel;
+  private final BlockingQueue<QueuedCommand> queue;
+  private final AtomicBoolean scheduled = new AtomicBoolean();
+
+  /**
+   * ArrayDeque to copy queue into when flushing in event loop.
+   */
+  private final ArrayDeque<QueuedCommand> writeChunk =
+      new ArrayDeque<QueuedCommand>(DEQUE_CHUNK_SIZE);
+
+  public AbstractWriteQueue(Channel channel) {
+    this.channel = Preconditions.checkNotNull(channel, "channel");
+    queue = new LinkedBlockingQueue<QueuedCommand>();
+  }
+
+  /**
+   * Schedule a flush on the channel.
+   */
+  void scheduleFlush() {
+    if (scheduled.compareAndSet(false, true)) {
+      // Add the queue to the tail of the event loop so writes will be executed immediately
+      // inside the event loop. Note DO NOT do channel.write outside the event loop as
+      // it will not wake up immediately without a flush.
+      channel.eventLoop().execute(later);
+    }
+  }
+
+  /**
+   * Enqueue a write command on the channel.
+   *
+   * @param command a write to be executed on the channel.
+   * @param flush true if a flush of the write should be schedule, false if a later call to
+   *              enqueue will schedule the flush.
+   */
+  void enqueue(Object command, boolean flush) {
+    enqueue(command, null, flush);
+  }
+
+  /**
+   * Enqueue a write command with a completion listener.
+   *
+   * @param command a write to be executed on the channel.
+   * @param listener to be notified on the completion of the write.
+   * @param flush true if a flush of the write should be schedule, false if a later call to
+   *              enqueue will schedule the flush.
+   */
+  void enqueue(Object command, ChannelFutureListener listener, boolean flush) {
+    queue.add(new QueuedCommand(command, listener));
+    if (flush) {
+      scheduleFlush();
+    }
+  }
+
+  /**
+   * Process the queue of commands and dispatch them to the stream. This method is only
+   * called in the event loop
+   */
+  void flush(ChannelHandlerContext ctx, ChannelPromise promise) {
+    try {
+      boolean flushed = false;
+      while (queue.drainTo(writeChunk, DEQUE_CHUNK_SIZE) > 0) {
+        while (writeChunk.size() > 0) {
+          writeChunk.poll().execute(ctx);
+        }
+        // Flush each chunk so we are releasing buffers periodically. In theory this loop
+        // might never end as new events are continuously added to the queue, if we never
+        // flushed in that case we would be guaranteed to OOM.
+        flushed = true;
```

Comment:
Since it is theoretically possible to get stuck writing forever, that would mean we never read anything.  We should probably create a bug to investigate ways to ensure we don't get stuck in this scenario.